TotalSeqB_CELLRANGER

Our data was sequenced using TotalSeqB chemistry and therefore cell
surface antibodies were used to multiplexed data. In our case 4 HTOs
(antibody barcodes) were used for multiplexing samples in a batch of
three. Totally 30 samples were sequenced in 10 lines using the
combination of these 4 HTOs.

The data we received came in the following format:

![](media/image14.png){width="6.267716535433071in"
height="2.0694444444444446in"}

We received 8 files per sample. Four GEX files with expression data and
4 Abs files with antibody information.

Furthermore, we received the following Excel sheet with HTO information
to demultiplex the data. Meaning for Sample MMC-OO79 HTO1 was used.
Which is the first one in Name column.

![](media/image13.png){width="5.109375546806649in"
height="4.06084208223972in"}

The experimental design was as follows:

![](media/image5.png){width="6.267716535433071in"
height="2.7083333333333335in"}

We learned there are two different ways of demultiplexing/analyzing this
data.

[First option: Use cellranger multi command]{.underline}

![](media/image9.png){width="6.267716535433071in"
height="3.4305555555555554in"}

This will be run through the cellranger_multi.py script.

![](media/image11.png){width="6.267716535433071in"
height="1.3055555555555556in"}

To run this we need in our working directory the following
CMO_reference, R1, R2, I1 and I2 files for each sample and a config file
for each sample.

A CMO_reference is analogous to the Samplesheet.csv used in standard
Illumina Sequencing and contains the barcode information for
demultiplexing. This file works in conjunction with a configuration file
(see the screenshot below) which contains the mapping between the sample
and the barcodes (HTO1, HTO2, HTO3, HTO4) along with the reference
genome and the library information. The HTO information was received
from the sequencing company.

![](media/image7.png){width="6.267716535433071in"
height="1.0694444444444444in"}

As well as the CMO_reference we need a csv config file with information
regarding the location of reference transcriptome location,
CMO_reference locations, I1,I2,R1 and R2 files location and finally the
link between HTO (barcodes) and samples. One example of config file is
the following. We need to create these config files for each fastq (
containing 3 samples). The sample ids were generated by the project
coordinator but can be any name.

![](media/image8.png){width="6.267716535433071in"
height="2.1805555555555554in"}

Example can be found on:

/share/data/RNA_Seq/MECFS_scRNAseq/Working/April19_2022/Cellranger

![](media/image10.png){width="6.267716535433071in"
height="1.0972222222222223in"}

**Use of script and working command will be:**

Note that cellranger will check for ids just before \_Ab or \_GEX.

/share/data/software/cellranger/cellranger-6.1.2/bin/cellranger multi
\--id=MMC-125-135-144-C \--csv=MMC-125-135-144-C.config

The id generates an output directory containing all of the
demultiplexing data and output explanation can be found in here:
[[https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/summary-multi]{.underline}](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/summary-multi)

![](media/image17.png){width="6.267716535433071in"
height="6.263888888888889in"}

![](media/image15.png){width="6.267716535433071in" height="3.0in"}

If I go to per_sample_outs and then to one of the sample folders:

![](media/image16.png){width="6.267716535433071in"
height="2.5972222222222223in"}

![](media/image12.png){width="6.267716535433071in"
height="1.2916666666666667in"}

Files for this analysis can be found in cellranger_multi folder

Following scRNASeq scripts have been developed with this output.
Information about the following steps are in:

[[https://drive.google.com/drive/folders/1fGJgqbToiJtsH9wwXl4z2l9Z2TdzD_55]{.underline}](https://drive.google.com/drive/folders/1fGJgqbToiJtsH9wwXl4z2l9Z2TdzD_55)

[Second option: Use cellranger count command and then run
Seurat]{.underline}

For the same sample format we have received there is a second way of
analyzing this data. This will be done by cellranger_count_Abs.py script

![](media/image6.png){width="6.267716535433071in"
height="1.2361111111111112in"}

[[https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis\\]{.underline}](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/feature-bc-analysis%5C)

Following this vignette we will follow this example.

![](media/image1.png){width="6.267716535433071in"
height="1.2638888888888888in"}

As previously we need to create two csv files: CMO reference and config
file with fastq path, sample name and library type. The difference with
the first option relies in the feature_type for this CMO reference file.

![](media/image3.png){width="6.267716535433071in"
height="1.3194444444444444in"}

While the config file for each fastq file needs to be as follows:

![](media/image4.png){width="6.267716535433071in" height="0.625in"}

Example can be found in:

![](media/image2.png){width="6.267716535433071in"
height="0.4583333333333333in"}

**Use of script and working command will be:**

The output file os this run will be different from the first Option.
Here there will be only one matrix for the three sequenced together
samples. Therefore we will need an additional step run with a R script.

Using the the three files in output
[[filtered_feature_bc_matrix]{.underline}](https://portal-us.medgenome.com/P2003609_04062022/CITESEQ/MMC-97-107-117-B_results_CITESEQ/filtered_feature_bc_matrix/)
folder we can run HTOdemult from Seurat:

[[https://satijalab.org/seurat/archive/v3.1/hashing_vignette.html]{.underline}](https://satijalab.org/seurat/archive/v3.1/hashing_vignette.html)

We load the three matrix files and follow the R script.

The R script for demultiplexing (HTO_demult.html) as well as examples of
CMO and config file can be found in cellranger_count_Abs folder
